File: packages/raisin-client-java/src/main/java/com/raisindb/client/operations/Archetypes.java
package com.raisindb.client.operations;

import com.raisindb.client.protocol.RequestContext;
import com.raisindb.client.protocol.RequestType;

import java.util.*;
import java.util.concurrent.CompletableFuture;

/**
 * Archetype management operations.
 */
public class Archetypes {

    private final Database database;

    public Archetypes(Database database) {
        this.database = database;
    }

    /**
     * Create a new Archetype.
     *
     * @param name      Archetype name
     * @param archetype Archetype definition
     * @return CompletableFuture with created Archetype
     */
    public CompletableFuture<Object> create(String name, Map<String, Object> archetype) {
        RequestContext context = database.getContext(null, null);
        Map<String, Object> payload = new HashMap<>();
        payload.put("name", name);
        payload.put("archetype", archetype);

        return database.getClient().getRequestTracker()
                .sendRequest(RequestType.ARCHETYPE_CREATE, context, payload);
    }

    /**
     * Get an Archetype by name.
     *
     * @param name Archetype name
     * @return CompletableFuture with Archetype
     */
    public CompletableFuture<Object> get(String name) {
        RequestContext context = database.getContext(null, null);
        Map<String, Object> payload = new HashMap<>();
        payload.put("name", name);

        return database.getClient().getRequestTracker()
                .sendRequest(RequestType.ARCHETYPE_GET, context, payload);
    }

    /**
     * List all Archetypes.
     *
     * @param publishedOnly If true, list only published Archetypes
     * @return CompletableFuture with list of Archetypes
     */
    public CompletableFuture<List<Object>> list(boolean publishedOnly) {
        RequestContext context = database.getContext(null, null);
        Map<String, Object> payload = new HashMap<>();
        payload.put("published_only", publishedOnly);

        return database.getClient().getRequestTracker()
                .sendRequest(RequestType.ARCHETYPE_LIST, context, payload)
                .thenApply(result -> {
                    if (result instanceof List) {
                        return (List<Object>) result;
                    }
                    return Collections.emptyList();
                });
    }

    /**
     * List all Archetypes (published and unpublished).
     *
     * @return CompletableFuture with list of Archetypes
     */
    public CompletableFuture<List<Object>> list() {
        return list(false);
    }

    /**
     * Update an Archetype.
     *
     * @param name      Archetype name
     * @param archetype Updated Archetype definition
     * @return CompletableFuture with updated Archetype
     */
    public CompletableFuture<Object> update(String name, Map<String, Object> archetype) {
        RequestContext context = database.getContext(null, null);
        Map<String, Object> payload = new HashMap<>();
        payload.put("name", name);
        payload.put("archetype", archetype);

        return database.getClient().getRequestTracker()
                .sendRequest(RequestType.ARCHETYPE_UPDATE, context, payload);
    }

    /**
     * Delete an Archetype.
     *
     * @param name Archetype name
     * @return CompletableFuture with result
     */
    public CompletableFuture<Object> delete(String name) {
        RequestContext context = database.getContext(null, null);
        Map<String, Object> payload = new HashMap<>();
        payload.put("name", name);

        return database.getClient().getRequestTracker()
                .sendRequest(RequestType.ARCHETYPE_DELETE, context, payload);
    }

    /**
     * Publish an Archetype.
     *
     * @param name Archetype name
     * @return CompletableFuture with result
     */
    public CompletableFuture<Object> publish(String name) {
        RequestContext context = database.getContext(null, null);
        Map<String, Object> payload = new HashMap<>();
        payload.put("name", name);

        return database.getClient().getRequestTracker()
                .sendRequest(RequestType.ARCHETYPE_PUBLISH, context, payload);
    }

    /**
     * Unpublish an Archetype.
     *
     * @param name Archetype name
     * @return CompletableFuture with result
     */
    public CompletableFuture<Object> unpublish(String name) {
        RequestContext context = database.getContext(null, null);
        Map<String, Object> payload = new HashMap<>();
        payload.put("name", name);

        return database.getClient().getRequestTracker()
                .sendRequest(RequestType.ARCHETYPE_UNPUBLISH, context, payload);
    }
}

File: packages/raisin-client-js/src/archetypes.ts
/**
 * Archetype management operations
 */

import { RequestContext, RequestType } from './protocol';

export class Archetypes {
  private sendRequest: (payload: unknown, requestType: string) => Promise<unknown>;

  constructor(
    _context: RequestContext,
    sendRequest: (payload: unknown, requestType: string) => Promise<unknown>
  ) {
    this.sendRequest = sendRequest;
  }

  /**
   * Create a new Archetype
   */
  async create(name: string, archetype: Record<string, unknown>): Promise<unknown> {
    return this.sendRequest(
      {
        name,
        archetype
      },
      RequestType.ArchetypeCreate
    );
  }

  /**
   * Get an Archetype by name
   */
  async get(name: string): Promise<unknown> {
    return this.sendRequest(
      {
        name
      },
      RequestType.ArchetypeGet
    );
  }

  /**
   * List all Archetypes
   */
  async list(publishedOnly = false): Promise<unknown[]> {
    const result = await this.sendRequest(
      {
        published_only: publishedOnly
      },
      RequestType.ArchetypeList
    );
    return Array.isArray(result) ? result : [];
  }

  /**
   * Update an Archetype
   */
  async update(name: string, archetype: Record<string, unknown>): Promise<unknown> {
    return this.sendRequest(
      {
        name,
        archetype
      },
      RequestType.ArchetypeUpdate
    );
  }

  /**
   * Delete an Archetype
   */
  async delete(name: string): Promise<unknown> {
    return this.sendRequest(
      {
        name
      },
      RequestType.ArchetypeDelete
    );
  }

  /**
   * Publish an Archetype
   */
  async publish(name: string): Promise<unknown> {
    return this.sendRequest(
      {
        name
      },
      RequestType.ArchetypePublish
    );
  }

  /**
   * Unpublish an Archetype
   */
  async unpublish(name: string): Promise<unknown> {
    return this.sendRequest(
      {
        name
      },
      RequestType.ArchetypeUnpublish
    );
  }
}
